Fix GC holes for all platforms and a bug for System V OSs in the codegen

LLITCHEV opened the pull request:
for GS cookie comparison on amd64.

Fixes issue #3382.

There is a bug in the genEmitGSCookieCheck method for amd64 having to do
with potentially trashing a second rgister (RDX) for a 2 register returned
struct.

It also sets the byref state for RAX, if needed. And GCRef and BrRef for
RDX, if needed.

pgavlin commented:
@dotnet/jit-contrib PTAL

CarolEidt commented:
LGTM

briansull commented on src/jit/codegenxarch.cpp:
```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+        else
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+        {
+            if (compiler->info.compRetType == TYP_REF)
```
This should be written this way:

if  (varTypeIsGC(compiler->info.compRetType))
{
    gcInfo.gcRegGCrefSetCur |= RBM_INTRET; 
}

CarolEidt replied:
@briansull - are you saying that these can never contain TYP_BYREFs that should go into gcRegByrefSetCur?

briansull replied:
Og waut, Sorry I didn't see the difference gcRegByrefSetCur vs gcRegGCrefSetCur 

So I will retract this comment

briansull replied:
This should be written this way:

gcInfo.gcMarkRegPtrVal(REG_INTRET, compiler->info.compRetType);

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->info.retStructDesc.passedInRegisters)
```
If checks on line 182 and 184 can be replaced by

if (compMethodReturnsMultiRegRetType()) ?

Note that this method is not handling structs returned in two registers.  In case of structs returned in a single registers, return type is normalized.

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+            if (compiler->info.retStructDesc.passedInRegisters)
+            {
+                assert(compiler->info.retStructDesc.eightByteCount > 0 &&
+                       compiler->info.retStructDesc.eightByteCount <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
```
This assert also needs modification if we change the if check above

briansull commented on src/jit/codegenxarch.cpp:
```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+                {
+                    var_types eightByteType = compiler->getEightByteType(compiler->info.retStructDesc, i);
+
+                    if (eightByteType == TYP_REF)
```
This should be written this way:

if (varTypeIsGC(eightByteType) {
if (i==0) {  gcInfo.gcRegGCrefSetCur |= RBM_INTRET; }
else  {  gcInfo.gcRegGCrefSetCur |= RBM_INTRET_1; }
}
with appropriate whitespace.
You can also use the (expr) ? val1 : val2 
but I am not a fan of that, 
someone may want to put a breakpoint on the RBM_INTRET_1 
to find out what method hit this in mscorlib etc... 
and they will have to find another way of doing that.

briansull commented:
LGTM after suggested changes

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+                    {
+                        gcInfo.gcRegByrefSetCur |= ((i == 0) ? RBM_INTRET : RBM_INTRET_1);
+                    }
+                }
```
The above for-loop can be replaced with to use existing routines genStructTypeOffset() and genStructReturnRegisters().  I would suggest to leverage existing routines rather than duplicating the code to compute return registers.

getStructTypeOffset(structDes, &type0, &type1, &offset0, &offset1);

if (varTypeIsGC(type0) || varTypeIsGC(type1))
{
     genStructReturnRegisters(type0, type1, &retReg0, &retReg1);

     if (type0 == TYP_REF)
         gcInfo.gcRegGCrefSetCur |= genRegMask(retReg0);
     else if (type0 == TYP_BYREF)
         gcInfo.gcRegByrefSetCur |= genRegMask(retReg0);

     if (type1 == TYP_REF)
         gcInfo.gcRegGCrefSetCur |= genRegMask(retReg1);
     else if (type1 == TYP_BYREF)
         gcInfo.gcRegByrefSetCur |= genRegMask(retReg1);         
}

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -188,7 +224,13 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+            regGSCheck = REG_RSI;
+#else // !FEATURE_UNIX_AMD64_STRUCT_PASSING
             regGSCheck = REG_RDX;
+#endif // !FEATURE_UNIX_AMD64_STRUCT_PASSING
```
We can replace lines 227-233 with

regGSCheck = REG_ARG_1;

Also we should replace REG_ECX with REG_ARG_0.

sivarv commented on src/jit/compiler.cpp:
```diff
@@ -4930,6 +4930,14 @@ int           Compiler::compCompileHelper (CORINFO_MODULE_HANDLE            clas
+        {
+            eeGetSystemVAmd64PassStructInRegisterDescriptor(info.compMethodInfo->args.retTypeClass, &(info.retStructDesc));
+        }
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
```
If the only use of this structDesc is in emitting GS Cookie check for a GT_RETURN node, can't this be computing on demand using eeGetSystemVAmd64PassStructInRegisterDescriptor(info.compMethodInfo->args.retTypeClass, &structDesc)?

IMO we are unnecessarily paying the cost of computing structDesc and storing it without knowing whether the method would need it.

sivarv commented on src/jit/codegenxarch.cpp:
```diff
@@ -188,7 +224,13 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
         if (compiler->lvaKeepAliveAndReportThis() && compiler->lvaTable[compiler->info.compThisArg].lvRegister &&
             (compiler->lvaTable[compiler->info.compThisArg].lvRegNum == REG_ECX))
         {
```
Even this check needs modification since thisPtr the first arg on Linux will be in REG_EDI.  Here we should replace REG_ECX with REG_ARG_0

sivarv replied:
Actually there is no trashing of RDX here by accident on Linux.  Since thisArg on Linux will be passed in REG_EDI the if check will always be false.  Hence gs check was using RCX which is REG_ARG3.  Since there is no bug in emitting GS cookie check please update the title of PR and also the bug filed.

sivarv replied:
Ok. Lubo offline pointed out that there could be a bug if thisptr is moved from edi to ecx in method prolog and kept alive throughout the method and such a method is returning a struct containing gc-refs in two return registers.

briansull commented on src/jit/codegenxarch.cpp:
```diff
@@ -176,8 +176,44 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+
+                for (unsigned i = 0; i < compiler->info.retStructDesc.eightByteCount; i++)
+                {
+                    var_types eightByteType = compiler->getEightByteType(compiler->info.retStructDesc, i);
```
This can be written this way:

if (i=0) { gcInfo.gcMarkRegPtrVal(REG_INTRET, eightByteType); }
else { gcInfo.gcMarkRegPtrVal(REG_INTRET_1, eightByteType); }

LLITCHEV commented:
Got expected conflicts here after the PR #3351 was merged. replacing this PR with https://github.com/dotnet/coreclr/pull/3459/files (or #3459 ), which also addresses the comments from this one.
